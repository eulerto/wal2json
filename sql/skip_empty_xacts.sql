-- predictability
SET synchronous_commit = on;

SELECT 'init' FROM pg_create_logical_replication_slot('regression_slot', 'wal2json');

CREATE TABLE xact_test(data text);
INSERT INTO xact_test VALUES ('before-test');

BEGIN;
-- perform operation in xact that creates and logs xid, but isn't decoded
SELECT * FROM xact_test FOR UPDATE;
SAVEPOINT foo;
-- and now actually insert in subxact, xid is expected to be known
INSERT INTO xact_test VALUES ('after-assignment');
COMMIT;
-- and now show those changes
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-xids', '0', 'skip-empty-xacts', '1');

BEGIN;
-- first insert
INSERT INTO xact_test VALUES ('main-txn');
SAVEPOINT foo;
-- now perform operation in subxact that creates and logs xid, but isn't decoded
SELECT 1 FROM xact_test FOR UPDATE LIMIT 1;
COMMIT;
-- and now show those changes
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'include-xids', '0', 'skip-empty-xacts', '1');

DROP TABLE xact_test;

SELECT pg_drop_replication_slot('regression_slot');