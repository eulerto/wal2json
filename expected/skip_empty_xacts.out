\set VERBOSITY terse
-- predictability
SET synchronous_commit = on;
DROP TABLE IF EXISTS tbl;
CREATE TABLE tbl (id int);
DROP VIEW IF EXISTS tbl_view;
NOTICE:  view "tbl_view" does not exist, skipping
CREATE MATERIALIZED VIEW tbl_view AS select * from tbl;
SELECT 'init' FROM pg_create_logical_replication_slot('regression_slot', 'wal2json');
 ?column? 
----------
 init
(1 row)

-- Now slot should have zero changes
SELECT count(*) = 0 FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL);
 ?column? 
----------
 t
(1 row)

-- Refreshing Materialized Views generates empty transactions
REFRESH MATERIALIZED VIEW tbl_view;
-- Now slot should have one change
SELECT count(*) = 1 FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL);
 ?column? 
----------
 t
(1 row)

REFRESH MATERIALIZED VIEW tbl_view;
-- The plugin should ignore the empty transaction
SELECT count(*) = 0 FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'skip-empty-xacts', '1');
 ?column? 
----------
 t
(1 row)

REFRESH MATERIALIZED VIEW tbl_view;
-- Writing transactions in chunks should work too
SELECT count(*) = 0 FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 
    'skip-empty-xacts', '1', 'write-in-chunks', '1');
 ?column? 
----------
 t
(1 row)

REFRESH MATERIALIZED VIEW tbl_view;
-- Write in chunks still works
SELECT count(*) = 2 FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'write-in-chunks', '1');
 ?column? 
----------
 t
(1 row)

SELECT 'stop' FROM pg_drop_replication_slot('regression_slot');
 ?column? 
----------
 stop
(1 row)

